synedrion/src/paillier/ring_pedersen.rs

```diff
@@ -1,134 +1,166 @@
+/// Implements the Definition 3.3 from the CGGMP'21 paper and related operations.
 use core::ops::Mul;
 
-use crypto_bigint::{PowBoundedExp, Square};
+use crypto_bigint::{Monty, NonZero, RandomMod, ShrVartime};
 use rand_core::CryptoRngCore;
 use secrecy::{ExposeSecret, SecretBox};
 use serde::{Deserialize, Serialize};
 
-use super::{PaillierParams, PublicKeyPaillierPrecomputed, SecretKeyPaillierPrecomputed};
-use crate::uint::{Bounded, Exponentiable, Retrieve, Signed, ToMontgomery};
+use super::{
+    rsa::{PublicModulus, PublicModulusWire, SecretPrimes, SecretPrimesWire},
+    PaillierParams,
+};
+use crate::{
+    tools::Secret,
+    uint::{Bounded, Exponentiable, Retrieve, Signed, ToMontgomery},
+};
 
-pub(crate) struct RPSecret<P: PaillierParams>(Bounded<P::Uint>);
+/// Ring-Pedersen secret.
+#[derive(Debug, Clone)]
+pub(crate) struct RPSecret<P: PaillierParams> {
+    primes: SecretPrimes<P>,
+    lambda: Secret<Bounded<P::Uint>>,
+}
 
 impl<P: PaillierParams> RPSecret<P> {
-    pub fn random(rng: &mut impl CryptoRngCore, sk: &SecretKeyPaillierPrecomputed<P>) -> Self {
-        // The random value will be reduced modulo `phi(N)` implicitly
-        // when used as an exponent modulo N later.
-        // So we are sampling it from this range to begin with.
-        Self(sk.random_field_elem(rng))
+    pub fn random(rng: &mut impl CryptoRngCore) -> Self {
+        let primes = SecretPrimesWire::<P>::random_safe(rng).into_precomputed();
+
+        let bound = SecretBox::init_with(|| {
+            NonZero::new(primes.totient().expose_secret().wrapping_shr_vartime(2))
+                .expect("totient / 4 is still non-zero because p, q >= 5")
+        });
+        let lambda = SecretBox::init_with(|| {
+            Bounded::new(P::Uint::random_mod(rng, bound.expose_secret()), P::MODULUS_BITS - 2)
+                .expect("totient < N < 2^MODULUS_BITS, so totient / 4 < 2^(MODULUS_BITS - 2)")
+        })
+        .into();
+
+        Self { primes, lambda }
+    }
+
+    pub fn lambda(&self) -> &SecretBox<Bounded<P::Uint>> {
+        &self.lambda
```

Comment:
So this surprised me at first, but then I realized there's an `AsRef` impl. I honestly think it'd be more readable if you extracted the `SecretBox` from the `Secret` explicitly.

Reply:
That's being done in #163